bootstrap/src/stream.mli

```diff
@@ -0,0 +1,89 @@
+(** Lazy-evaluated immutable {!type:stream}.
+
+    The {!type:stream} type is an immutable stack-like container.  Operations on
+    the {!type:stream} are amortized with lazy execution and (with a few
+    exceptions) are O(1).
+*)
+
+open Rudiments
+
+(** {1 Type and derivations} *)
+
+type 'a elm =
+  | Nil
+  | Cons of 'a * 'a t
+and 'a t = 'a elm Lazy.t
+
+include Formattable_intf.S_poly with type 'a t := 'a t
+
+(** {1 Creation} *)
+
+val empty: 'a t
+(** Return an empty {!type:stream}. *)
+
+val init: usize -> f:(usize -> 'a) -> 'a t
+(** Initialize stream.  [init len ~f:(fun i -> ...)] lazily initializes a stream
+    of given length, where [f] provides the value for each element at given
+    index. *)
+
+(** {1 Length} *)
+
+val length: 'a t -> usize
+(** Return stream length.  This forces and counts every node and is
+ * Ω(n)/O(n). *)
+
+val is_empty: 'a t -> bool
+(** Return [true] if stream length is 0; [false] otherwise. *)
+
+(** {1 Element access} *)
+
+val hd: 'a t -> 'a
+(** Return head (first) stream element, or halt if stream is empty. *)
+
+val tl: 'a t -> 'a t
+(** Return tail (a stream of all elements except head), or halt if list is
+    empty. *)
+
+(** {1 Combining and partitioning} *)
+
+val push: 'a -> 'a t -> 'a t
+(** Push element onto stream and return resulting stream with element as hd. *)
+
+val pop: 'a t -> 'a * 'a t
+(** Pop head element off stream and return the decomposed element and remainder
+    stream.  Halt if the input list is empty. *)
+
+val concat: 'a t -> 'a t -> 'a t
+(** Concatenate two streams. *)
+
+val split: 'a t -> usize -> 'a t * 'a t
+(** Split the stream with elements [\[0..len)] into streams with elements
+    [\[0..n)] and [\[n..len)].  If stream contains fewer than [n] elements,
+    returns a stream with elements [\[0..len)] and and empty stream.  Split is
+    O(1), but forcing the first element of the second returned stream is
+    Ω(n)/O(n). *)
```

Comment:
Realized that the split and rev_split cases are actually `Ω(1)/O(n)` since we could have done it in a way that creates no monolith work in one of the streams. For example
```
let t = init 1_000_000_000 ~f:(fun i -> i) in
let _, t1 = split t 0 in
let _ = hd t1
```
Since we split at 0, no extra work is needed to access the first element of the second list since we didn't have to walk over any elements. This is the best-case time.

I'll just change the comment to `O(n)`.